receiver/wavefrontreceiver/wavefront_parser.go

```diff
@@ -0,0 +1,283 @@
+	// This comes from SignalFx Gateway code that has the capability to
+	// remove CollectD tags from the name of the metric.
+	var toAddDims map[string]string
+	index := strings.Index(metricName, "..")
+	for {
+		metricName, toAddDims = labelsFromName(&metricName)
+		if len(toAddDims) == 0 {
+			if index == -1 {
+				metricName = strings.Replace(metricName, "..", ".", -1)
+			}
+
+			break
+		}
+
+		for k, v := range toAddDims {
+			labelKeys = append(labelKeys, &metricspb.LabelKey{Key: k})
+			labelValues = append(labelValues, &metricspb.LabelValue{
+				Value:    v,
+				HasValue: true,
+			})
+		}
+	}
+	return metricName, labelKeys, labelValues
+}
+
+func buildLabels(tags string) (keys []*metricspb.LabelKey, values []*metricspb.LabelValue, err error) {
+	if tags == "" {
+		return
+	}
+	for {
+		parts := strings.SplitN(tags, "=", 2)
+		if len(parts) != 2 {
+			return nil, nil, fmt.Errorf("failed to break key for [%s]", tags)
+		}
+
+		key := parts[0]
+		rest := parts[1]
+		tagLen := len(key) + 1 // Length of key plus separator and yet to be determined length of the value.
+		var value string
+		if len(rest) > 1 && rest[0] == '"' {
+			// Skip until non-escaped double quote.
+			foundEscape := false
+			i := 1
+			for ; i < len(rest); i++ {
+				if rest[i] != '"' && rest[i] != 'n' {
+					continue
+				}
+				isPrevCharEscape := rest[i-1] == '\\'
+				if rest[i] == '"' && !isPrevCharEscape {
+					// Non-escaped double-quote, it is the end of the value.
+					break
+				}
+				foundEscape = foundEscape || isPrevCharEscape
+			}
+
+			value = rest[1:i]
+			tagLen += len(value) + 2 // plus 2 to account for the double-quotes.
+			if foundEscape {
+				// Per implementation of Wavefront SDK only double-quotes and
+				// newline characters are escaped. See the link below:
+				// https://github.com/wavefrontHQ/wavefront-sdk-go/blob/2c5891318fcd83c35c93bba2b411640495473333/senders/formatter.go#L20
+				value = escapedCharReplacer.Replace(value)
+			}
+		} else {
+			// Skip until space.
+			i := 0
+			for ; i < len(rest) && rest[i] != ' '; i++ {
+			}
+			value = rest[:i]
+			tagLen += i
+		}
+
+		keys = append(keys, &metricspb.LabelKey{Key: key})
+		values = append(values, &metricspb.LabelValue{
+			Value:    value,
+			HasValue: true})
+
+		tags = strings.TrimLeft(tags[tagLen:], " ")
+		if tags == "" {
+			break
+		}
+	}
+
+	return
+}
+
+func unDoubleQuote(s string) string {
+	n := len(s)
+	if n < 2 {
+		return s
+	}
+
+	if s[0] == '"' && s[n-1] == '"' {
+		return s[1 : n-1]
+	}
+	return s
+}
+
+// labelsFromName tries to pull out dimensions out of name in the format name[k=v,f=x]-morename
+// would return name-morename and extract dimensions (k,v) and (f,x)
+// if we encounter something we don't expect use original name.
+// This is a bit complicated to avoid allocations, string.split allocates, while slices
+// inside same function, do not.
+// TODO: This is being copied from collectdreceiver, expose it and call it from that package.
```

Comment:
I remember seeing a PR that did extract this function or make it public. Shouldn't we be using that?

Reply:
I started that one when I saw the need in this one :) I will update this now since the other was already merged.